### ISSUE TITLE
check that readers passed in are valid when they are passed in

### ISSUE THREAD

**Issue by ericblade** [ericblade/quagga2#124]
I'm not exactly sure how that would work, but see https://github.com/serratus/quaggaJS/issues/360#issuecomment-579480829

**Comment by danikaze** [comment]
JavaScript is a non-typed language, and if you start with that approach, to be consistent, why don't type check EVERYTHING? that's a bit of overkill in my opinion...

So, I'm not sure that this feature is needed, since that's extra runtime check, while you can use TypeScript to solve that problem...

**Comment by ericblade** [comment]
Well, there's a specific list of valid readers, and if you pass in something that isn't valid at init time, then sometime later, you get an error that doesn't appear to be at all related to what you passed init.  That's a pretty bad experience, IMO.  That list also could be a dynamic list, with the support I've added for external reader modules .. and once that is working fully, I'd like to remove the built-in readers to separate modules, so that we don't have to ship the code for all readers in the bundle, and it'll also prevent people from making the common mistake of enabling *all* readers.  

I think it would be good to throw in init, or at least warn in init, then ignore the bad reader, so it doesn't trigger errors at a place that is completely irrelevant to what the error is.

**Comment by kevinpapst** [comment]
[This list](https://github.com/ericblade/quagga2/blob/master/src/decoder/barcode_decoder.js#L17)? I'd would throw a hard error in init() if an unknown reader is configured.

**Comment by ericblade** [comment]
Yep that's the one.  I had filed this when I ran across the problem, just for future reference, I hadn't really thought about it at the time ...

But now that we're discussing it :-)  I'm wondering if it should match the readers in init() into the list, subtract any that are not valid, then throw if there's no valid left -- or if it should just throw if you pass in any invalid ones.  

Method 1: If you pass in multiple readers, but only 1 of them is invalid, the others work, without it breaking the entire thing.  Downside is that if you don't check the log, you have no idea why the invalid reader isn't working.

Method 2: It just straight up fails, and you fix your broken init() :-)

I'm leaning towards #2 because this is a library that is targeted towards devs, rather than an end-user application, where i'd want to try to figure out what the user meant if possible.

**Comment by kevinpapst** [comment]
I would totally go with method 2. 
Better show the dev directly the mistake instead of "hiding" it. 

Will that work with dynamic reader list? I assume that those have to be registered upfront somehow and that you can then check against ... I don't know, some kind of identifier?

**Comment by ericblade** [comment]
Yeah, it should.  Dynamically added readers don't seem to be working in LiveStream mode, so that's a problem (and also quite difficult to write a test for, I think, but it might be possible that ImageStream can be handled and used to substitute for it..) so there might be some changes in how those get registered.. but right now, you call registerReader before you call init() and it should add it to the list.  

I think once I'm done converting them all to typescript, I'm going to take the UPC reader as it's 100% known functionality, and test it as a plugin.

**Comment by danikaze** [comment]
What I was saying is, if you want to write consistent code, why would you **only** check the reader list? Why stop there and don't check the rest of the configuration?
Checking only the readers but not the rest looks like a half-ass solution for me.

So, if we want to have a consistent library, there are two options:
- Check the whole configuration
- Check nothing and rely on other ways (i.e. check on compilation time with TypeScript)

The problem of the first approach is that you require more code and more execution (because it's done in run time) and that's an overhead, which might not be useful for everyone (i.e. for us using typescript).

It would be something similar to [prop-types](https://www.npmjs.com/package/prop-types) in React, which I used to use a lot, but I stopped using it on favor of TypeScript strong-typed projects.

Also one thing to consider, is having different builds for production and development, so the checks are only done in development, without including any code in production. That's also a valid approach which doesn't add overhead in production. And if you don't use test in development you are doing it just wrong :)

**Comment by kevinpapst** [comment]
Not all people build quagga themselves. They just use the pre-built dist.
That might sound totally wrong to the full-hearted frontend dev, but is reality. 
They will likely not switch between dev and prod built.

Additionally: wrong spelled configuration shouldn't be a problem during runtime, while a misspelling in the readers name is.

> ... i just had another thought, that it might even be a better idea to change init() so that readers is BarcodeReader[] rather than string[] .. does that seem weird?

Not at all and I could work with both. But to be devils advocate here: this is a library and you should think of the users that are not frontend devs. If there will be a solution to still drop the prebuilt file(s) into the assets directory and use this new style in `init()` without setting up a complete built-stack then it is "only" a massive BC break. 
My wish: please don't force anyone into needing a mandatory frontend built-environment. 

Had a quick look at initReaders() and the constructors of the readers. Half of them take `opts` and `supplements` as arguments. Can those params be moved  easily to the userland?

**Comment by ericblade** [comment]
I think the crux of this problem, is that passing invalid reader strings generates a runtime error -- after the fact -- that doesn't point to where the actual problem is.  Screwing up the configuration in most other ways, probably just ends up with an incorrectly configured setup.  That is the difference between doing a runtime validation on the readers versus the rest of the configuration.    I guess I'd have to go and write an init() with a bunch of incorrect parameters to test that theory, but in either case -- the library shouldn't blow up in a way that points to code inside it failing, if it's userland that's wrong.  

There's already this setup that i inherited, that readers can be either string[] or object[] where object includes the name, the opts, and the supplements.  Using a BarcodeReader class instead of a string name for those could be a new option, that would both be easily checkable (!instanceof BarcodeReader) and might not require any pre-configuration (Quagga.registerReader before Quagga.init)

This is future, anyway, I think my priorities when I do have time to work on this right now are:

TypeScript 
Automated Testing
Re-implement Workers
